packages/lab/src/stepped-tracker/stepReducer.ts

```diff
@@ -0,0 +1,170 @@
+import type { Step } from "./Step";
+import { assignSteps, autoStageSteps, flattenSteps, resetSteps } from "./utils";
+
+export namespace StepReducer {
+  export interface State {
+    steps: Step.Record[];
+    flatSteps: Step.Record[];
+    activeStep: Step.Record | null;
+    previousStep: Step.Record | null;
+    nextStep: Step.Record | null;
+    activeStepIndex: number;
+    started: boolean;
+    ended: boolean;
+  }
+
+  export type Action =
+    | { type: "next" }
+    | { type: "previous" }
+    | { type: "error" }
+    | { type: "warning" }
+    | { type: "clear" };
+
+  export type Options = {
+    activeStepId?: string;
+  };
+}
+
+/**
+ * Extracts all step ids from a array of steps,
+ * including top and lower levels.
+ *
+ * I can't use this type because our codebase
+ * is running below TypeScript 5, where "const"
+ * in function generics was firstly introduced.
+ * Leaving this here for future reference.
+ * */
+type AllowedActiveStepIds<
+  S extends Step.Record[],
+  Acc extends string = never,
+> = S extends [{ id: infer ID; substeps?: infer SS }, ...infer R]
+  ? ID extends string
+    ? R extends Step.Record[]
+      ? SS extends Step.Record[]
+        ? AllowedActiveStepIds<R, Acc | AllowedActiveStepIds<SS>>
+        : AllowedActiveStepIds<R, Acc | ID>
+      : Acc
+    : Acc
+  : Acc;
+
+export default function stepReducer(
```

Comment:
Is there any known use-case where the Reducer may need to be extended to support additional state or actions ?
james may have visibility of that 

if we are giving someone a hook it might be something that we consider adding if there is interest